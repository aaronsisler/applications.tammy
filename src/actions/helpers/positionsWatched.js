export const addPositionWatch = (positionWatched) => ({
    type: 'ADD_POSITION_WATCH',
    positionWatched,
});

export const removePositionWatch = (positionId) => ({
    type: 'REMOVE_POSITION_WATCH',
    positionId,
});

export const setPositionsWatched = (positionsWatched) => ({
    type: 'SET_POSITIONS_WATCHED',
    positionsWatched,
});

export const setPositionWatchLevel = (positionId, notificationLevel) => ({
    type: 'SET_POSITION_WATCH_LEVEL',
    positionId,
    notificationLevel,
});
