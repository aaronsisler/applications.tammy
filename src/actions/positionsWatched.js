import database from 'Firebase/firebase';
import {
    addSubscription,
    removeSubscription,
    setPositionsWatched,
    setSubscriptionLevel
} from 'Actions/helpers/positionsWatched';

export const startAddSubscription = (positionId, subscriptionLevel) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).update({ subscriptionLevel })
        .then(() => dispatch(addSubscription(positionId, subscriptionLevel)));
}

export const startRemoveSubscription = (positionId) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).remove()
        .then(() => dispatch(removeSubscription(positionId)));
}

export const startSetPositionsWatched = () => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`position_watch/${userId}`).once('value').then((snapshot) => {
        const positionsWatched = [];
        snapshot.forEach((childSnapshot) => {
            positionsWatched.push({
                positionId: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        return dispatch(setPositionsWatched(positionsWatched));
    });
}

export const startSetSubscriptionLevel = (positionId, subscriptionLevel) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).update({ subscriptionLevel })
        .then(() => dispatch(setSubscriptionLevel(positionId, subscriptionLevel)));
}



