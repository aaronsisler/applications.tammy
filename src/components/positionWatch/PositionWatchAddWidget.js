import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
    startAddPositionWatch,
    startRemovePositionWatch,
    startSetPositionWatchLevel
} from 'Actions/positionsWatched';

export class PositionWatchAddWidget extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            notificationLevel: 'NONE',
        }
    }

    handleSetPositionWatchLevel = (e) => {
        const notificationLevel = e.target.value;
        if (this.state.notificationLevel === 'NONE') {
            this.props.startAddPositionWatch(this.props.positionId, notificationLevel);
        } else if (notificationLevel === 'NONE') {
            this.props.startRemovePositionWatch(this.props.positionId);
        } else {
            this.props.startSetPositionWatchLevel(this.props.positionId, notificationLevel);
        }
        return this.setState(() => ({ notificationLevel }));
    }

    render() {
        return (
            <div className="position_watch_add_widget">
                <div className="position_watch_add_widget__select_wrapper">
                    <div className="position_watch_add_widget__title">
                        Notification&nbsp;Level:
                    </div>
                    <div className="position_watch_add_widget__select">
                        <select
                            className="select"
                            value={this.state.notificationLevel}
                            onChange={this.handleSetPositionWatchLevel}
                        >
                            <option value='ALL'>ALL</option>
                            <option value='SOME'>SOME</option>
                            <option value='REQUIRED'>REQUIRED</option>
                            <option value='NONE'>NONE</option>
                        </select>
                    </div>
                </div>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = (state) => ({
    positionId: state.position.positionId,
});

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
    startAddPositionWatch: (positionId, notificationLevel) => dispatch(startAddPositionWatch(positionId, notificationLevel)),
    startRemovePositionWatch: (positionId) => dispatch(startRemovePositionWatch(positionId)),
    startSetPositionWatchLevel: (positionId, notificationLevel) => dispatch(startSetPositionWatchLevel(positionId, notificationLevel)),
});

export default connect(mapStateToProps, mapDispatchToProps)(PositionWatchAddWidget);

PositionWatchAddWidget.propTypes = {
    positionId: PropTypes.string.isRequired,
    startAddPositionWatch: PropTypes.func.isRequired,
    startRemovePositionWatch: PropTypes.func.isRequired,
    startSetPositionWatchLevel: PropTypes.func.isRequired,
};
