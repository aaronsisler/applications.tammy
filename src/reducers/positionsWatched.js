const positionsWatchedReducerDefaultState = [];

const positionsWatchedReducer = (state = positionsWatchedReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_POSITIONS_WATCHED':
            return action.positionsWatched;
        case 'ADD_POSITION_WATCH':
                return [...state, action.positionWatched];
        case 'REMOVE_POSITION_WATCH':
            return state.filter((positionwatched) => positionwatched.positionId != action.positionId);
        case 'SET_POSITION_WATCH_LEVEL':
            return state.map((positionWatched) => {
                if (positionWatched.positionId === action.positionId) {
                    return {
                        ...positionWatched,
                        notificationLevel: action.notificationLevel
                    }
                }
                return positionWatched;
            })
        default:
            return state;
    }
};

export default positionsWatchedReducer;

