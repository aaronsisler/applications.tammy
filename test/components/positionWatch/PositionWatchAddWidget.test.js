import React from 'react';
import { shallow } from 'enzyme';
import { PositionWatchAddWidget } from 'PositionWatch/PositionWatchAddWidget';
import { positionsWatchedStore } from '../../fixtures/positionsWatched';

describe('PositionWatchAddWidget', () => {
    const [positionWatched] = positionsWatchedStore;
    const { positionId } = positionWatched;
    const startAddPositionWatch = jest.fn();
    const startRemovePositionWatch = jest.fn();
    const startSetPositionWatchLevel = jest.fn();
    let wrapper;

    const buildWrapper = () => {
        wrapper = shallow(
            <PositionWatchAddWidget
                positionId={positionId}
                startAddPositionWatch={startAddPositionWatch}
                startRemovePositionWatch={startRemovePositionWatch}
                startSetPositionWatchLevel={startSetPositionWatchLevel}
            />
        );
    };

    beforeEach(() => {
        buildWrapper();
    });

    it('should render PositionWatchAddWidget correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should set the notificationLevel correctly on construction', () => {
        expect(wrapper.find('select').prop('value')).toEqual('NONE');
    });

    describe('when the notification select input is changed', () => {
        describe('when the current notification level is NONE', () => {
            const newNotificationLevel = 'ALL';
            const event = { target: { value: newNotificationLevel } };

            it('should call startAddPositionWatch with positionId and new notification level', () => {
                wrapper.find('select').simulate('change', event);

                expect(startAddPositionWatch).toHaveBeenLastCalledWith(positionId, newNotificationLevel);
            });

            it('should set the new notification level correctly in the select input', () => {
                wrapper.find('select').simulate('change', event);

                expect(wrapper.find('select').prop('value')).toEqual(newNotificationLevel);
            });

            it('should set the new notification level correctly in the state', () => {
                wrapper.find('select').simulate('change', event);

                expect(wrapper.state('notificationLevel')).toEqual(newNotificationLevel);
            });
        });

        describe('when the notification level is changed to NONE', () => {
            const newNotificationLevel = 'NONE';
            const event = { target: { value: newNotificationLevel } };

            beforeEach(() => {
                wrapper.setState({ notificationLevel: 'ALL' });
                expect(wrapper.find('select').prop('value')).toEqual('ALL');
            });

            it('should call startRemovePositionWatch with positionId', () => {
                wrapper.find('select').simulate('change', event);

                expect(startRemovePositionWatch).toHaveBeenLastCalledWith(positionId);
            });

            it('should set the new notification level correctly in the select input', () => {
                wrapper.find('select').simulate('change', event);

                expect(wrapper.find('select').prop('value')).toEqual(newNotificationLevel);
            });

            it('should set the new notification level correctly in the state', () => {
                wrapper.find('select').simulate('change', event);

                expect(wrapper.state('notificationLevel')).toEqual(newNotificationLevel);
            });
        });

        describe('when the notification level is changed to something other than NONE', () => {
            const newNotificationLevel = 'SOME';
            const event = { target: { value: newNotificationLevel } };

            beforeEach(() => {
                wrapper.setState({ notificationLevel: 'ALL' });
                expect(wrapper.find('select').prop('value')).toEqual('ALL');
            });

            it('should call startSetPositionWatchLevel with positionId and new notification level', () => {
                wrapper.find('select').simulate('change', event);

                expect(startSetPositionWatchLevel).toHaveBeenLastCalledWith(positionId, newNotificationLevel);
            });

            it('should set the new notification level correctly in the select input', () => {
                wrapper.find('select').simulate('change', event);

                expect(wrapper.find('select').prop('value')).toEqual(newNotificationLevel);
            });

            it('should set the new notification level correctly in the state', () => {
                wrapper.find('select').simulate('change', event);

                expect(wrapper.state('notificationLevel')).toEqual(newNotificationLevel);
            });
        });
    });
});
