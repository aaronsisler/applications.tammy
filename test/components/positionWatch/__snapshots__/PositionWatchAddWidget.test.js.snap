// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionWatchAddWidget should render PositionWatchAddWidget correctly 1`] = `
<div
  className="position_watch_add_widget"
>
  <div
    className="position_watch_add_widget__select_wrapper"
  >
    <div
      className="position_watch_add_widget__title"
    >
      NotificationÂ Level:
    </div>
    <div
      className="position_watch_add_widget__select"
    >
      <select
        className="select"
        onChange={[Function]}
        value="NONE"
      >
        <option
          value="ALL"
        >
          ALL
        </option>
        <option
          value="SOME"
        >
          SOME
        </option>
        <option
          value="REQUIRED"
        >
          REQUIRED
        </option>
        <option
          value="NONE"
        >
          NONE
        </option>
      </select>
    </div>
  </div>
</div>
`;
